### ISSUE TITLE
 [Suggestion] Different travis builds method at each PR

### ISSUE THREAD

**Issue by A-Hilaly** [DestructHub/ProjectEuler#135]
Hey !

Since pull requesting any changes make` stats.py `/ `stats.exs` run all the solutions for all the languages. I suggest to make these build independent from "already builded solutions" and only build new suggested solutions. Also make running all the solutions a possible option (when stats.py is modified for example).

My idea is to try to use destruct.github.io as a data source for ProjectEuler repo since it is hosted 24/7 and we have control of it.
If destruct.github.io contains a list of solutions and configurations (as hidden frame,  REST page or visible frame ( well presented for visitors too)), we can parse and make sure `stats.py` / `stats.exs` make the right job 

- Building and validate news problems 

- Building news solutions to already existing problems.

- Build all solutions & problems ( actual default build )

- A combination of the listed build ( sync / Async )

**Comment by ryukinix** [comment]
First, thanks to bringing this subject to here. This is really a old thing which I really want improve.

Yes, we already thought about a different (and better) approach to build the tests and I agreed that the current is not good enough. But the idea in these days that I proposed was about handling the diff commits on the master branch and the PR branch and getting a list of files changed, just on the travis using only `git` operations. I'm still think that this will be the ideal solution, as services like `codecov` implements to get differential coverage on each PR.

Unfortunately, we do not thought on a proper solution to implement that in this way.

For me, I'm really fine to change the current approach for building the tests, but I don't like to add the a new complexity relying to human maintenance on a list of solutions implemented. I think this is prone error. I prefer avoid that.

What you think about the differential `git` operation? I think hacking a little about the so many options and operations on `git` we can do that and I think this is more reliable and more trusty too. Since we just try making a current solution better (added to the website stuff as solved) this will skiped, but not in git approach because the file was changed.

If you still prefer the website approach, try elaborate more the pros we can have with it instead of the git approach. We really need discuss that.

God call: @rafaelcn @lubien

**Comment by ryukinix** [comment]
One thing that I forgot, I liked the idea of a static website being good for visitors, ok? What I don't like is to rely that for the tests.

**Comment by lubien** [comment]
1. About test a single solution

As @ryukinix pointed, we've been thinking about this for a while. Either locally and/or on the CI this could be useful a lot.

2. About the CI testing only modified solutions

There's no arguing here, this is for the best. Working with some git diff and travis maybe we can get this working out someway. Requires some testing from our part.

3. About the site.

I'm really confused. I couldn't really understand your idea and before I get any conclusions on what to think about it can you explain with more details?

**Comment by A-Hilaly** [comment]
@ryukinix I understand that's a bit complicated to implement all this and make a hole project build depend on a web site content. It was just my way to avoid using git manipulations and I was wondering how do i tell stats.py that this last builded problem is "validated" and you don't have to build it next time (validation is done when someone confirm that someone solution is correct, not only build success). I like the idea of git differentials especially codecov if we can target and get the name and type of files added/modified to decide which build to run.

@lubien 
I am sorry my idea wasn't so clear. What I mean is using `destruct.github.io` like sort of API or Data center (source) that holds number of problems already solved and validated, build configurations (default build, forced build, a series of builds etc); So each time the build runs, `stats.py` use request lib to parse for example `destruct.github.Io/xxxx/build_config` and `destruct.github.Io/xxxx/validated_problems`, and make build decision regarding these configurations. We can control the builds many times as we can without modifying the PR ( Change configuration at destruct.github.io and rebuild test for the same PR).

**Comment by A-Hilaly** [comment]
@ryukinix @lubien I agree with the idea that git differentials and codecov are the best choice here, I just say when trying to make different builds on travis without opening a new PR this could be useful !

**Comment by ryukinix** [comment]
I think are you really right about changing this behavior (building whole project on each PR is cumbersome). We'll hold that a little to discuss more, but I still really prefer the git differential approach. 

To validate the solutions the hashs is not enough? Well... an user can add too a wrong hash and this is hard to catch.

**Comment by lubien** [comment]
@A-Hilaly indeed it would require some work. Not only building a microservice but hosting it too. 

For a GitHub repo that relies on CI from Travis I think using a microservice is overkilling it so unfortunately I'm going to disagree to your proposal.

I'd go for @ryukinix method of diffing code and testing modified/added solutions. Of course it has drawbacks like we are supposing all previous tests are passing and only the current ones are error prone which could later stab us in the back maybe.

**Comment by lubien** [comment]
I vote to close this issue and start a new one on the subject of diff-based CI builds.

**Comment by A-Hilaly** [comment]
> To validate the solutions the hashs is not enough? Well... an user can add too a wrong hash and this is hard to catch.

Exactly. That's the "validation" thats worries me the most.
> Of course it has drawbacks like we are supposing all previous tests are passing and only the current ones are error prone which could later stab us in the back maybe.

It is true that it can be dangerous to dare building only recent uploads/modifications. On the other hand, I totally agree that git differentials are necessary anyways to catch what was modified or added (even in my suggestion), the hosted service was more like to solve what @ryukinix said about wrong solutions (hashes).
I think we all agree on using differentials now. Ill close this issue and create a new one after learning more about using git diffs.
